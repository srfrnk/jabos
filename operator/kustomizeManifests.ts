import { Request, Response, NextFunction } from 'express';

import genericManifests from './genericManifests'
import { CustomizeRequest, FinalizeRequest, SyncRequest } from './metaControllerHooks';
import { getRepo } from './misc';

export default {
  async sync(syncRequest: Request, response: Response, next: NextFunction) {
    const request: SyncRequest = syncRequest.body;
    genericManifests.debugRequest('kustomize', 'sync', request.object, request);

    const repo = getRepo(request);
    const latestCommit = repo.status.latestCommit;
    const spec: any = request.object.spec;
    const namespace = request.object.metadata.namespace;

    // Have to filter, due to the following bug: https://github.com/metacontroller/metacontroller/issues/414
    const dockerImages = Object.values(request.related["DockerImage.jabos.io/v1"] || {}).filter((r: any) => r.metadata.namespace === namespace);
    if (dockerImages.length !== spec.dockerImages.length) {
      // Have to reject the sync to maintain idempotent response - due to: https://github.com/metacontroller/metacontroller/issues/414
      throw new Error(`No DockerImages from same namespace.`);
    }

    const replacementPrefix = spec.replacementPrefix;
    const replacementSuffix = spec.replacementSuffix;
    const replacements = spec.replacements;

    await genericManifests.sync({
      metricName: 'kustomize',
      type: 'kustomize',
      metricLabel: 'kustomize',
      env: {
        'COMMIT': latestCommit,
        'REPLACEMENT_STRINGS': Object.entries(
          replacements).map(replacement => `s/${replacementPrefix}${replacement[0]}${replacementSuffix}/${replacement[1]}/g`
          ).join(';'),
        'IMAGES': JSON.stringify(dockerImages.map((image: any) => ({ "name": image.spec.imageName, "newTag": latestCommit })))
      },
    }, request, response);
  },

  async customize(customizeRequest: Request, response: Response, next: NextFunction) {
    const request: CustomizeRequest = customizeRequest.body;
    genericManifests.debugRequest('kustomize', 'customize', request.parent, request);

    await genericManifests.customize('kustomize', request, response, [{
      apiVersion: "jabos.io/v1",
      resource: "docker-images",
      namespace: request.parent.metadata.namespace,
      names: request.parent.spec.dockerImages
    }]);
  },

  async finalize(finalizeRequest: Request, response: Response, next: NextFunction) {
    const request: FinalizeRequest = finalizeRequest.body;
    genericManifests.debugRequest('kustomize', 'finalize', request.object, request);

    await genericManifests.finalize('kustomize', request, response);
  }
}
