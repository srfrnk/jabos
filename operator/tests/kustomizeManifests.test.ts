import { Request, Response } from 'express';
import mockResponse from './mockResponse';
import kustomizeManifests from '../kustomizeManifests';
import { setMock, clearMock } from './settingsMock';

beforeEach(() => {
  setMock();
});

afterEach(() => {
  clearMock();
});

test('kustomizeManifests sync', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        kind: 'object_kind',
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          uid: 'uid_value'
        },
        spec: {
          commitTLAKey: 'commitTLAKey_value',
          targetNamespace: 'namespace_value',
          gitRepository: 'gitRepository_value',
          path: '.',
          dockerImages: [],
          replacementPrefix: 'replacementPrefix_value',
          replacementSuffix: 'replacementSuffix_value',
          replacements: {}
        }
      },
      related: {
        'GitRepository.jabos.io/v1': [
          {
            metadata: {
              namespace: 'namespace_value'
            },
            spec: {
              url: 'url_value',
              branch: 'branch_value'
            },
            status: {
              latestCommit: 'commit_value'
            }
          }
        ]
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await kustomizeManifests.sync(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('kustomizeManifests sync no image found', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        kind: 'object_kind',
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          uid: 'uid_value'
        },
        spec: {
          commitTLAKey: 'commitTLAKey_value',
          targetNamespace: 'namespace_value',
          gitRepository: 'gitRepository_value',
          path: '.',
          dockerImages: ['invalid_image'],
          replacementPrefix: 'replacementPrefix_value',
          replacementSuffix: 'replacementSuffix_value',
          replacements: {}
        }
      },
      related: {
        'GitRepository.jabos.io/v1': [
          {
            metadata: {
              namespace: 'namespace_value'
            },
            spec: {
              url: 'url_value',
              branch: 'branch_value'
            },
            status: {
              latestCommit: 'commit_value'
            }
          }
        ]
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await expect(kustomizeManifests.sync(req, res as unknown as Response, () => { /* NOOP */ })).rejects.toMatchSnapshot();
});

test('kustomizeManifests customize', async () => {
  const req = {
    body: {
      parent: {
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value'
        },
        spec: {
          gitRepository: 'gitRepository_value',
          dockerImages: [],
        }
      },
    }
  } as unknown as Request;
  const res = mockResponse();
  await kustomizeManifests.customize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('kustomizeManifests finalize leave', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          annotations: {}
        },
        spec: {
          cleanupPolicy: 'Leave'
        }
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await kustomizeManifests.finalize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('kustomizeManifests finalize delete', async () => {
  const req = {
    body: {
      controller: {
        metadata: {
          name: 'controller_name'
        }
      },
      object: {
        kind: 'object_kind',
        metadata: {
          name: 'name_value',
          namespace: 'namespace_value',
          annotations: { deployedManifest: 'deployedManifest_value' },
          uid: 'object_uid'
        },
        spec: {
          cleanupPolicy: 'Delete',
          targetNamespace: 'namespace_value',
        }
      },
      attachments: {
        'Job.batch/v1': []
      }
    }
  } as unknown as Request;
  const res = mockResponse();
  await kustomizeManifests.finalize(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0]).toMatchSnapshot();
  expect(res.status).toHaveBeenCalledWith(200);
});

test('Issue #60', async () => {
  const req = {
    body: require('./test_input/issue#60.2.json')
  } as Request;

  const res = mockResponse();
  await kustomizeManifests.sync(req, res as unknown as Response, () => { /* NOOP */ });
  expect(res.json.mock.calls[0][0].attachments[2].metadata.annotations['metacontroller.k8s.io/last-applied-configuration']).toBeUndefined();
  expect(res.status).toHaveBeenCalledWith(200);
});
