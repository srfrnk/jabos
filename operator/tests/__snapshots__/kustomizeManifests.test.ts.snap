// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kustomizeManifests customize 1`] = `
Array [
  Object {
    "relatedResources": Array [
      Object {
        "apiVersion": "jabos.io/v1",
        "names": Array [
          "gitRepository_value",
        ],
        "namespace": "namespace_value",
        "resource": "git-repositories",
      },
      Object {
        "apiVersion": "jabos.io/v1",
        "names": Array [],
        "namespace": "namespace_value",
        "resource": "docker-images",
      },
    ],
  },
]
`;

exports[`kustomizeManifests finalize delete 1`] = `
Array [
  Object {
    "annotations": Object {},
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "*",
            ],
            "resources": Array [
              "*",
            ],
            "verbs": Array [
              "delete",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "v1",
        "kind": "ServiceAccount",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "cleaner-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "cleaner-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "cleaner-name_value",
            "namespace": "namespace_value",
          },
        ],
      },
      Object {
        "apiVersion": "batch/v1",
        "kind": "Job",
        "metadata": Object {
          "labels": Object {
            "type": "manifest-cleaner",
          },
          "name": "manifest-clean-name_value",
        },
        "spec": Object {
          "activeDeadlineSeconds": 1234,
          "backoffLimit": 1234,
          "completionMode": "NonIndexed",
          "completions": 1,
          "parallelism": 1,
          "template": Object {
            "metadata": Object {
              "annotations": Object {
                "manifests": "deployedManifest_value",
              },
              "labels": Object {
                "builder": "manifest-clean-name_value",
              },
              "name": "manifest-clean-name_value",
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "object_uid",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixmanifest-cleaner:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "manifest-cleaner",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
              "securityContext": Object {
                "runAsNonRoot": true,
              },
              "serviceAccountName": "cleaner-name_value",
              "volumes": Array [
                Object {
                  "emptyDir": Object {},
                  "name": "temp",
                },
                Object {
                  "downwardAPI": Object {
                    "items": Array [
                      Object {
                        "fieldRef": Object {
                          "fieldPath": "metadata.annotations['manifests']",
                        },
                        "path": "manifests.tar.gz.b64",
                      },
                    ],
                  },
                  "name": "manifests",
                },
              ],
            },
          },
          "ttlSecondsAfterFinished": 1234,
        },
      },
    ],
    "finalized": false,
  },
]
`;

exports[`kustomizeManifests finalize leave 1`] = `
Array [
  Object {
    "annotations": Object {},
    "attachments": Array [],
    "finalized": true,
  },
]
`;

exports[`kustomizeManifests sync 1`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "Role",
        "metadata": Object {
          "name": "deployer-name_value",
          "namespace": "namespace_value",
        },
        "rules": Array [
          Object {
            "apiGroups": Array [
              "*",
            ],
            "resources": Array [
              "*",
            ],
            "verbs": Array [
              "get",
              "list",
              "watch",
              "create",
              "update",
              "patch",
            ],
          },
        ],
      },
      Object {
        "apiVersion": "rbac.authorization.k8s.io/v1",
        "kind": "RoleBinding",
        "metadata": Object {
          "name": "deployer-name_value",
          "namespace": "namespace_value",
        },
        "roleRef": Object {
          "apiGroup": "rbac.authorization.k8s.io",
          "kind": "Role",
          "name": "deployer-name_value",
        },
        "subjects": Array [
          Object {
            "kind": "ServiceAccount",
            "name": "builder-gitRepository_value",
            "namespace": "namespace_value",
          },
        ],
      },
      Object {
        "apiVersion": "batch/v1",
        "kind": "Job",
        "metadata": Object {
          "labels": Object {
            "manifest-builder-type": "kustomize-manifests",
            "type": "manifest-builder",
          },
          "name": "manifest-name_value-commit_value",
        },
        "spec": Object {
          "activeDeadlineSeconds": 1234,
          "backoffLimit": 1234,
          "completionMode": "NonIndexed",
          "completions": 1,
          "parallelism": 1,
          "template": Object {
            "metadata": Object {
              "labels": Object {
                "builder": "manifest-name_value-commit_value",
              },
              "name": "manifest-name_value-commit_value",
            },
            "spec": Object {
              "containers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "TYPE",
                      "value": "kustomize-manifests",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "COMMIT",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "COMMIT_LABEL",
                      "value": "deployedCommit",
                    },
                    Object {
                      "name": "METRIC_NAME",
                      "value": "kustomizeManifestsBuilder",
                    },
                    Object {
                      "name": "METRIC_LABELS",
                      "value": "{\\"namespace\\":\\"namespace_value\\",\\"kustomize_manifests\\":\\"name_value\\"}",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixpost-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "post-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/timer",
                      "name": "timer",
                      "readOnly": true,
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                  ],
                },
              ],
              "initContainers": Array [
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "URL",
                      "value": "url_value",
                    },
                    Object {
                      "name": "BRANCH",
                      "value": "branch_value",
                    },
                    Object {
                      "name": "COMMIT",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "METRIC_NAME",
                      "value": "kustomizeManifestsBuilder",
                    },
                    Object {
                      "name": "METRIC_LABELS",
                      "value": "{\\"namespace\\":\\"namespace_value\\",\\"kustomize_manifests\\":\\"name_value\\"}",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                    Object {
                      "name": "JABOS_OPERATOR_URL",
                      "value": "http://operator.settings_namespace:1234/",
                    },
                  ],
                  "image": "settings_imagePrefixpre-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "pre-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/gitTemp",
                      "name": "git-temp",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/timer",
                      "name": "timer",
                    },
                  ],
                },
                Object {
                  "args": Array [],
                  "env": Array [
                    Object {
                      "name": "SRC_PATH",
                      "value": ".",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "COMMIT",
                      "value": "commit_value",
                    },
                    Object {
                      "name": "REPLACEMENT_STRINGS",
                      "value": "",
                    },
                    Object {
                      "name": "IMAGES",
                      "value": "[]",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                  ],
                  "image": "settings_imagePrefixkustomize-manifest-builder:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "kustomize-manifest-builder",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/gitTemp",
                      "name": "git-temp",
                      "readOnly": true,
                    },
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/build",
                      "name": "build",
                    },
                  ],
                },
                Object {
                  "args": Array [
                    "namespace_value",
                    "kustomize-manifests",
                    "name_value",
                  ],
                  "env": Array [
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "TARGET_NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "TYPE",
                      "value": "kustomize-manifests",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "KIND",
                      "value": "object_kind",
                    },
                    Object {
                      "name": "CONTROLLER",
                      "value": "controller_name",
                    },
                    Object {
                      "name": "NAMESPACE",
                      "value": "namespace_value",
                    },
                    Object {
                      "name": "NAME",
                      "value": "name_value",
                    },
                    Object {
                      "name": "OBJECT_UID",
                      "value": "uid_value",
                    },
                  ],
                  "image": "settings_imagePrefixmanifest-deployer:settings_build_number",
                  "imagePullPolicy": "Always",
                  "name": "manifest-deployer",
                  "resources": Object {
                    "limits": Object {
                      "cpu": "500m",
                      "memory": "500Mi",
                    },
                    "requests": Object {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": Object {
                    "allowPrivilegeEscalation": false,
                    "capabilities": Object {
                      "drop": Array [
                        "ALL",
                      ],
                    },
                    "readOnlyRootFilesystem": true,
                    "runAsNonRoot": true,
                  },
                  "stdin": true,
                  "tty": true,
                  "volumeMounts": Array [
                    Object {
                      "mountPath": "/manifests",
                      "name": "manifests",
                      "readOnly": true,
                    },
                    Object {
                      "mountPath": "/tmp",
                      "name": "temp",
                    },
                    Object {
                      "mountPath": "/build",
                      "name": "build",
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
              "securityContext": Object {
                "runAsNonRoot": true,
              },
              "serviceAccountName": "builder-gitRepository_value",
              "volumes": Array [
                Object {
                  "emptyDir": Object {},
                  "name": "manifests",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "git-temp",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "temp",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "build",
                },
                Object {
                  "emptyDir": Object {},
                  "name": "timer",
                },
              ],
            },
          },
          "ttlSecondsAfterFinished": 1234,
        },
      },
    ],
    "status": Object {
      "conditions": Array [
        Object {
          "status": "False",
          "type": "Synced",
        },
      ],
    },
  },
]
`;

exports[`kustomizeManifests sync no image found 1`] = `[Error: No DockerImages from same namespace.]`;
